src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
+
+namespace Microsoft.CodeAnalysis.IntroduceVariable
+{
+    internal abstract partial class AbstractIntroduceParameterService<
+        TExpressionSyntax,
+        TInvocationExpressionSyntax,
+        TIdentifierNameSyntax> : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+        where TInvocationExpressionSyntax : TExpressionSyntax
+        where TIdentifierNameSyntax : TExpressionSyntax
+    {
+        protected abstract bool IsContainedInParameterizedDeclaration(SyntaxNode node);
+        protected abstract SyntaxNode GenerateExpressionFromOptionalParameter(IParameterSymbol parameterSymbol);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, textSpan, cancellationToken) = context;
+            if (document.Project.Solution.Workspace.Kind == WorkspaceKind.MiscellaneousFiles)
+            {
+                return;
+            }
+
+            var action = await IntroduceParameterAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+            if (action != null)
+            {
+                context.RegisterRefactoring(action, textSpan);
+            }
+        }
+
+        public async Task<CodeAction?> IntroduceParameterAsync(
+            Document document, TextSpan textSpan, CancellationToken cancellationToken)
```

Comment:
why is this method public.  i'm curios why it's a method at all (and not just inlined into ComputeRefactoringsAsync).

Reply:
it was public bc it used to be called in a separate file at one point, but that file has since been removed. So should I just move all that logic up to compute refactoringsasync?

Reply:
that would be my pref.

Reply:
removed